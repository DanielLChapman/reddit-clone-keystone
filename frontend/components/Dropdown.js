import React, { useState } from 'react';
import PropTypes from 'prop-types'

function Dropdown(props) {
    /*
        let options = ['Hot', 'Best', 'New', 'Top', 'Rising'];
        const [dropValue, setDropValue] = useState('Hot');
        <Dropdown returnFunc={setDropValue} selected={dropValue} options={options} />    
    */
    return (
        <select value={props.selected} onChange={(e) => {
            props.returnFunc(e.target.value);
        }}>
            {
                props.options && props.options.map((x, i) => (
                    <option key={i} value={x}>{x}</option>
                ))
            }
        </select>
    );
}

Dropdown.propTypes = {
    returnFunc: PropTypes.func.isRequired,
    options: PropTypes.array.isRequired,
    selected: PropTypes.string.isRequired,
};
  

export default Dropdown;