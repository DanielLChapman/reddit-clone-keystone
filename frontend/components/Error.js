import React from 'react';

function Error(props) {
    return (
        <span>
            {props.error}
        </span>
    );
}

export default Error;