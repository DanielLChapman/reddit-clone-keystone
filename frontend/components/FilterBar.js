import React, { Component } from 'react';
import { FaHotjar, FaBahai, FaGopuram, FaChartLine} from 'react-icons/fa';



class FilterBar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            li1: true,
            li2: false,
            li3: false,
            li4: false
        }
    }

    toggleClass = (num) => {
        let state = {
            li1: false,
            li2: false,
            li3: false,
            li4: false
        };
        state['li'+num] = true;
        this.setState({...state});
    }

    render() {
        let style = {
            display: "none"
        };
        if (this.state.li3) {
            style= {
                display: "inline-block"
            };
        }
        return (
            <section className="filter-bar"> 
                <h5>Popular Posts</h5>
                <ul>
                    <li className={this.state.li1 ? 'active' : null} onClick={() => this.toggleClass(1)}><FaHotjar /> Hot</li>
                    <li className={this.state.li2 ? 'active' : null} onClick={() => this.toggleClass(2)}><FaBahai /> New</li>
                    <li className={this.state.li3 ? 'active' : null} onClick={() => this.toggleClass(3)}><FaGopuram /> Top</li>
                    <li className="time-selector-for-filter-bar" style={style}>Today</li>
                    <li className={this.state.li4 ? 'active' : null} onClick={() => this.toggleClass(4)}><FaChartLine /> Rising</li>
                </ul>
            </section>
        );
    }
}

export default FilterBar;