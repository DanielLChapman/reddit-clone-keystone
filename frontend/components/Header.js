import Link from 'next/link';
import Nav from './Nav';


export default function Header() {
  return (
    <>
      <div className="bar">
        <span>
          <Link href="/">App name</Link>
        </span>
        <Nav />
      </div>
    </>
  );
}
