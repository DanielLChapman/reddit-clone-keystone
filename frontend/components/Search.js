import { useLazyQuery } from '@apollo/client';
import { resetIdCounter, useCombobox } from 'downshift';
import gql from 'graphql-tag';
import debounce from 'lodash.debounce';
import { useRouter } from 'next/router';
import React, { useState } from 'react';
import styled from 'styled-components';
import capitalize from '../lib/capitalize';
import bestMatch from '../lib/getEditDistance';


const SEARCH_QUERY = gql`
    query SEARCH_QUERY($searchTerm: String!) {
        getSearchResults(
            searchTerm: $searchTerm,
        ) {
            id
            title
            subscriber {
                id
            }
            slug
            name
        },
    }
`;

function Search(props) {
    const router = useRouter();

    const [findItems, {loading, data, error}] = useLazyQuery(SEARCH_QUERY, {
        fetchPolicy: 'no-cache'
    });

    const findItemsDebounced = debounce(findItems, 350);

    resetIdCounter();

    const DropDown = styled.div`
        border:1px solid black;
        background: white;
        width: 91.5%;
        left: 38px;
        top:55px;
        position:absolute;
    `;

    const DropDownItem = styled.div`
        border-bottom: 1px solid lightgray;
        background: ${(props) => (props.highlighted ? '#f7f7f7' : 'white')};
        padding: .5rem;
        transition: all 0.2s;
        ${(props) => (props.highlighted ? 'padding-left: 2rem;' : null)};
        display: flex;
        align-items: center;
        font-size: .95rem;
        border-left: 10px solid
        ${(props) => (props.highlighted ? 'lightgray' : 'white')};
  `;

    let items = data?.getSearchResults || [];
    //sort by most accurate


    const { inputValue, getMenuProps, getInputProps, getComboboxProps,isOpen, highlightedIndex ,getItemProps, selectedItem} = useCombobox({
        items,
        onInputValueChange() {
            findItemsDebounced({
                variables: {
                    searchTerm: inputValue
                }
            });
            
        },
        onSelectedItemChange({selectedItem}) {
            router.push(`/r/${selectedItem.slug}`);
        },
        itemToString: (item) => item?.name || '',
    });

    

    return (
        <div className="search">
            <button type="submit" className="searchButton">
                <i className="fa fa-search"></i>
            </button>
            <div  {...getComboboxProps()} >
                <input {...getInputProps({
                    type: 'search',
                    placeholder: 'Search',
                    id: 'search',
                    className:'searchTerm'
                    
                })} 
                />
                {/*<input type="text" className="searchTerm" value={search} placeholder="Search" onChange={(e) => {
                    updateSearch(e.target.value);
                }} />*/}
            </div>
            
            <DropDown {...getMenuProps()}>
                {isOpen && 
                    items.map((item, index) => {

                        return (
                            <DropDownItem
                                {...getItemProps({ item, index })}
                                key={item.id}
                                highlighted={index === highlightedIndex}
                            >
                                {capitalize( item.slug)}
                            </DropDownItem>
                    )}
                        
                    )
                }
                {isOpen && !items.length && !loading && (
                    <DropDownItem>Sorry, No items found for {inputValue}</DropDownItem>
                )}
            </DropDown>
            
        </div>
    );
}

export default Search;