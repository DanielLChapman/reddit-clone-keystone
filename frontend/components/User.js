import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }
  }
  }
`;

export function useUser() {

  const { data } = useQuery(CURRENT_USER_QUERY);
  let test = {...data};

  
  if (test.authenticatedItem) {
    let x = Object.assign({}, test.authenticatedItem);
    //count up post votes
    let postVotes = 0;
    let commentVotes = 0;
    
    x.posts.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? postVotes += 1 : postVotes -= 1;
      })
    });
    //count up comment votes
    x.comments.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? commentVotes += 1 : commentVotes -= 1;
      })
    });

    x.postVotes = postVotes;
    x.commentVotes = commentVotes;

    return x;
  } else {
    return data?.authenticatedItem;
  }

  
}
