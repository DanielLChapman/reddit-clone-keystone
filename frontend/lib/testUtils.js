import casual from 'casual';

  
// seed it so we get consistent results
casual.seed(777);

const fakeUser = (overrides) => (
    {
    __typename: 'User',
    id: '4234',
    name: casual.name,
    username: casual.name,
    owner: [],
    email: casual.email,
    permissions: ['ADMIN'],
    moderating: [],
    subreddits: [],
    posts: [],
    postvotes: [],
    firstSignin: false,
    comments: [],
    commentvotes:[],
    ...overrides,
});

const fakeSubreddit = (overrides) => (
    {
        __typename: 'Subreddit',
        id: casual.uuid,
        name: casual.string,
        title: casual.string,
        slug: casual.string,
        sidebar: casual.string,
        rules: casual.string,
        description: casual.string,
        status: 'PUBLIC',
        ownder: fakeUser(),
        moderators: [],
        subscriber: [],
        posts: [],
        createdAt: Date.now(),
    }
);

const fakeLinkPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'www.google.com',
    type: 'Link',
    comments: [],
    removed: 'False',
    id: casual.uuid,


});

const fakeMediaPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'https://cdn.fileinfo.com/img/ss/sm/png_79.jpg',
    type: 'Link',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeTextPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    content: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: '',
    type: 'Text',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeComment = (overrides) => (
    {
    __typename: 'Comment',
    content: casual.sentence,
    user: fakeUser(),
    post: fakeTextPost(),
    createdAt: Date.now(),
    votes: [],
    parent: null,
    id: casual.uuid,
    children: [],

});


// Fake LocalStorage
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  }



export {

    LocalStorageMock,
    fakeLinkPost,
    fakeComment,
    fakeMediaPost,
    fakeSubreddit,
    fakeTextPost,
    fakeUser,

  };