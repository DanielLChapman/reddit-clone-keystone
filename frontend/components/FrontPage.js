import Link from 'next/link';
import React from 'react';
import { useUser } from './User';

function FrontPage(props) {
    const user = useUser();
    return (
        <div>
            {user && (
                <Link href="/subreddits/create">Create Your Own Subreddit</Link>
            )}
            
        </div>
    );
}

export default FrontPage;