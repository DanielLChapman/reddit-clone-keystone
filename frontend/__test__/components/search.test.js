import { render, screen } from '@testing-library/react';
import { fakeLinkPost, fakeMediaPost, fakeTextPost, fakeUser } from '../../lib/testUtils';
import {MockedProvider} from '@apollo/react-testing';
import { CURRENT_USER_QUERY } from '../../components/User';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import Search, { SEARCH_QUERY } from '../../components/Search';
import userEvent from '@testing-library/user-event';

let fakePosts = [fakeLinkPost(), fakeMediaPost(), fakeMediaPost(), fakeTextPost()];
let fakeSubreddits = fakePosts.map((x) => {
    return x.subreddit;
});

const mocks = [
    {
      // When someone requests this query and variable combo
      request: {
        query: SEARCH_QUERY,
        variables: {
            searchTerm: 'c'
        }
      },
      result: () => {
        return {
          data: {
            getSearchResults: [
                {   id: fakeSubreddits.id,
                    title: fakeSubreddits.title,
                    subscriber: [],
                    slug: fakeSubreddits.slug,
                    name: fakeSubreddits.name,
                },
            ]
          },
        }
      },
    },
    {
        request: {
            query: CURRENT_USER_QUERY,
        },
        result: {
            data: {
                
            }
        }
    }
]

jest.useFakeTimers("modern")

describe('It correctrly displays the search area with a dropdown', () => {
    it('renders the empty search bar', () => {
        const { container, debug } = render(
            <MockedProvider >
              <Search />
            </MockedProvider>
          );

          expect(container).toMatchSnapshot();
    });




    it('renders the search bar with mocks for lazyQuery', async () => {
        const { container, debug } = render(
            <MockedProvider mocks={mocks}>
              <Search />
            </MockedProvider>

          );

          await screen.findByTestId('search-bar-results');
    

          expect(container).toMatchSnapshot();
    })
})