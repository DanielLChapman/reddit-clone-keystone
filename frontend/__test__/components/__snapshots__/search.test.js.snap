// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It correctrly displays the search area with a dropdown renders the empty search bar 1`] = `
<div>
  <div
    class="search"
  >
    <button
      class="searchButton"
      type="submit"
    >
      <i
        class="fa fa-search"
      />
    </button>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-0-menu"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-0-menu"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="searchTerm"
        id="search"
        placeholder="Search"
        type="search"
        value=""
      />
    </div>
    <div
      aria-labelledby="downshift-0-label"
      class="Dropdown__DropDown-sc-1liqluw-0 TgajO dropdown-window dropdown-not-opened"
      data-testid="search-bar-results"
      id="downshift-0-menu"
      role="listbox"
    />
  </div>
</div>
`;

exports[`It correctrly displays the search area with a dropdown renders the search bar with mocks for lazyQuery 1`] = `
<div>
  <div
    class="search"
  >
    <button
      class="searchButton"
      type="submit"
    >
      <i
        class="fa fa-search"
      />
    </button>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="downshift-0-menu"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-0-menu"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="searchTerm"
        id="search"
        placeholder="Search"
        type="search"
        value=""
      />
    </div>
    <div
      aria-labelledby="downshift-0-label"
      class="Dropdown__DropDown-sc-1liqluw-0 TgajO dropdown-window dropdown-not-opened"
      data-testid="search-bar-results"
      id="downshift-0-menu"
      role="listbox"
    />
  </div>
</div>
`;
